src/txrequest.cpp

```diff
@@ -0,0 +1,722 @@
+public:
+    explicit PriorityComputer(bool deterministic) :
+        m_k0{deterministic ? 0 : GetRand(0xFFFFFFFFFFFFFFFF)},
+        m_k1{deterministic ? 0 : GetRand(0xFFFFFFFFFFFFFFFF)} {}
+
+    Priority operator()(const uint256& txhash, NodeId peer, bool preferred) const
+    {
+        uint64_t low_bits = CSipHasher(m_k0, m_k1).Write(txhash.begin(), txhash.size()).Write(peer).Finalize() >> 1;
+        return low_bits | uint64_t{preferred} << 63;
+    }
+
+    Priority operator()(const Announcement& ann) const
+    {
+        return operator()(ann.m_txhash, ann.m_peer, ann.m_preferred);
+    }
+};
+
+// Definitions for the 3 indexes used in the main data structure.
+//
+// Each index has a By* type to identify it, a By*View data type to represent the view of announcement it is sorted
+// by, and an By*ViewExtractor type to convert an announcement into the By*View type.
+// See https://www.boost.org/doc/libs/1_58_0/libs/multi_index/doc/reference/key_extraction.html#key_extractors
+// for more information about the key extraction concept.
+
+// The ByPeer index is sorted by (peer, state == CANDIDATE_BEST, txhash)
+//
+// Uses:
+// * Looking up existing announcements by peer/txhash, by checking both (peer, false, txhash) and
+//   (peer, true, txhash).
+// * Finding all CANDIDATE_BEST announcements for a given peer in GetRequestable.
+struct ByPeer {};
+using ByPeerView = std::tuple<NodeId, bool, const uint256&>;
+struct ByPeerViewExtractor
+{
+    using result_type = ByPeerView;
+    result_type operator()(const Announcement& ann) const
+    {
+        return ByPeerView{ann.m_peer, ann.m_state == State::CANDIDATE_BEST, ann.m_txhash};
+    }
+};
+
+// The ByTxHash index is sorted by (txhash, state, priority).
+//
+// Note: priority == 0 whenever state != CANDIDATE_READY.
+//
+// Uses:
+// * Deleting all announcements with a given txhash in ForgetTxHash.
+// * Finding the best CANDIDATE_READY to convert to CANDIDATE_BEST, when no other CANDIDATE_READY or REQUESTED
+//   announcement exists for that txhash.
+// * Determining when no more non-COMPLETED announcements for a given txhash exist, so the COMPLETED ones can be
+//   deleted.
+struct ByTxHash {};
+using ByTxHashView = std::tuple<const uint256&, State, Priority>;
+class ByTxHashViewExtractor {
+    const PriorityComputer& m_computer;
+public:
+    ByTxHashViewExtractor(const PriorityComputer& computer) : m_computer(computer) {}
+    using result_type = ByTxHashView;
+    result_type operator()(const Announcement& ann) const
+    {
+        const Priority prio = (ann.m_state == State::CANDIDATE_READY) ? m_computer(ann) : 0;
+        return ByTxHashView{ann.m_txhash, ann.m_state, prio};
+    }
+};
+
+enum class WaitState {
+    //! Used for announcements that need efficient testing of "is their timestamp in the future?".
+    FUTURE_EVENT,
+    //! Used for announcements whose timestamp is not relevant.
+    NO_EVENT,
+};
+
+WaitState GetWaitState(const Announcement& ann)
```

Comment:
The `WaitState` enum and this function seem superfluous now, since you can just replace the enum with a bool and use `IsWaiting()`.

Reply:
Gone.